trunk/spamdb/spamdb.8
.\"	$OpenBSD: spamdb.8,v 1.15 2007/05/31 19:20:29 jmc Exp $
.\"
.\" Copyright (c) 2004 Bob Beck.  All rights reserved.
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd September 11, 2007
.Dt SPAMDB 8
.Os
.Sh NAME
.Nm spamdb
.Nd spamd database tool
.Sh SYNOPSIS
.Nm spamdb
.Oo Fl D Oc
.Op Fl Y Ar synctarget
.Oo Oo Fl Tt Oc
.Fl a Ar keys Oc
.Oo Oo Fl Tt Oc
.Fl d Ar keys Oc
.Sh DESCRIPTION
.Nm
manipulates the spamd database in
.Pa /var/db/spamd
used for
.Xr spamd 8 .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl D
debug mode
.It Fl Y Ar synctarget
Add target
.Ar synctarget
to receive synchronisation messages.
.Ar synctarget
can be either an IPv4 address for unicast messages
or a network interface and optional TTL value for multicast messages
to the group 224.0.1.240.
If the multicast TTL is not specified, a default value of 1 is used.
This option can be specified multiple times.
See also
.Sx SYNCHRONISATION
below.
.It Fl a Ar keys
Add or update the entries for
.Ar keys .
This can be used to whitelist one or more IP addresses
(i.e. circumvent the greylisting process altogether)
by adding all IP addresses as keys to the spamd database for WHITE entries.
If any
.Ar keys
specified match entries already in the spamd database,
.Nm
updates the entry's time last seen to now.
.It Fl d Ar keys
Delete entries for
.Ar keys .
.It Fl T
Add or delete the keys as SPAMTRAP entries.
See the
.Sx GREYTRAPPING
section of
.Xr spamd 8
for more information.
Must be used in conjunction with the
.Fl a
or
.Fl d
option.
.It Fl t
Add or delete the keys as TRAPPED entries.
See the
.Sx GREYTRAPPING
section of
.Xr spamd 8
for more information.
Must be used in conjunction with the
.Fl a
or
.Fl d
option.
.El
.Pp
If adding or deleting a SPAMTRAP address
.Pq Fl T ,
.Ar keys
should be specified as email addresses:
.Pp
.Dl spamtrap@mydomain.org
.Pp
Otherwise
.Ar keys
must be numerical IP addresses.
.Ss DATABASE OUTPUT FORMAT
If invoked without any arguments,
.Nm
lists the contents of the database in a text format.
For SPAMTRAP entries the format is:
.Pp
.Dl type|mailaddress
.Pp
where
.Em type
will be SPAMTRAP and
.Em mailaddress
will be the email address for which any connections received by
.Xr spamd 8
will be blacklisted if mail is sent to this address.
.Pp
For TRAPPED entries the format is:
.Pp
.Dl type|ip|expire
.Pp
where
.Em type
will be TRAPPED,
.Em IP
will be the IP address blacklisted due to hitting a spamtrap, and
.Em expire
will be when the IP is due to be removed from the blacklist.
.Pp
For GREY entries, the format is:
.Pp
.Dl type|source IP|helo|from|to|first|pass|expire|block|pass
.Pp
For WHITE entries, the format is:
.Pp
.Dl type|source IP|||first|pass|expire|block|pass
.Pp
The fields are as follows:
.Pp
.Bl -tag -width "source IP" -offset indent -compact
.It type
.Em WHITE
if whitelisted or
.Em GREY
if greylisted
.It source IP
IP address the connection originated from
.It helo
what the connecting host sent as identification in the HELO/EHLO command in the
SMTP dialogue
.It from
envelope-from address for
.Em GREY
(empty for
.Em WHITE
entries)
.It to
envelope-to address for
.Em GREY
(empty for
.Em WHITE
entries)
.It first
time the entry was first seen
.It pass
time the entry passed from being
.Em GREY
to being
.Em WHITE
.It expire
time the entry will expire and be removed from the database
.It block
number of times a corresponding connection received a temporary
failure from
.Xr spamd 8
.It pass
number of times a corresponding connection has been seen to pass
to the real MTA by
.Xr spamlogd 8
.El
.Pp
Note that times are in seconds since the Epoch, in the manner returned by
.Xr time 3 .
Times may be converted to human readable format using:
.Pp
.Dl $ date -r <value>
.Pp
.Sh SYNCHRONISATION
.Nm
supports realtime synchronisation of white or blacklist states between
a number of
.Em spamd
daemons running on multiple machines,
using the
.Fl Y
option.
.Pp
The following example will send a unicast synchronisation messages.
.Bd -literal -offset indent
.Em whitelisting
# /usr/local/sbin/spamdb -Y foo.example.org -a 1.2.3.4 2.3.4.5
# /usr/local/sbin/spamdb -Y foo.example.org -Y bar.example.org -a 1.2.3.4
.Pp
.Em blacklisting
# /usr/local/sbin/spamdb -Y foo.example.org -ta 1.2.3.4 2.3.4.5
.Ed
.Pp
If the file
.Pa /usr/local/etc/spamd/spamd.key
exists,
.Nm
will calculate the message-digest fingerprint (checksum) for the file
and use it as a shared key to authenticate the synchronisation messages.
The file itself can contain any data.
For example, to create a secure random key:
.Bd -literal -offset indent
# dd if=/dev/random of=/usr/local/etc/spamd/spamd.key bs=2048 count=1
.Ed
.Pp
The IP send to peers is logged via
.Xr syslogd 8
at
.Dv LOG_INFO
level.
The following
.Xr syslog.conf 5
section can be used to log connection details to a dedicated file:
.Bd -literal -offset indent
!spamdb
daemon.info			/var/log/spamdb
.Ed
.Pp
The file needs to be copied to all hosts
sending or receiving synchronisation messages.
.Sh FILES
/var/db/spamd
.Pp
/usr/local/etc/spamd/spamd.key
.Sh SEE ALSO
.Xr spamd.conf 5 ,
.Xr spamd 8 ,
.Xr spamd-setup 8 ,
.Xr spamlogd 8
.Sh HISTORY
The
.Nm
command
appeared in
.Ox 3.5 .


trunk/spamdb/spamdb.c
/*	$OpenBSD: spamdb.c,v 1.23 2007/05/31 20:00:16 cnst Exp $	*/

/*
 * Copyright (c) 2004 Bob Beck.  All rights reserved.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <db.h>
#include <err.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef __FreeBSD__
#include <syslog.h>
#endif
#include <time.h>
#include <netdb.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>

#include "grey.h"
#ifdef __FreeBSD__
#include "sync.h"
#endif

/* things we may add/delete from the db */
#define WHITE 0
#define TRAPHIT 1
#define SPAMTRAP 2

#ifdef __OpenBSD__
struct syslog_data	 sdata	= SYSLOG_DATA_INIT;
#else
#define	syslog_r(l, s, args...)	syslog(l,args)
#define	openlog_r(i, l, f, s)	openlog(i, l, f)
#define	closelog_r(l)	        closelog()
int sdata = 0;					/* dummy */
#endif

int	dblist(DB *);
int	dbupdate(DB *, char *, int, int);

#ifdef __FreeBSD__
/* things we need for sync */
void 	check_opt_sequence(char **, int);

u_short sync_port;
int syncrecv;
int syncsend;
int debug = 0;
int greylist = 1;
FILE *grey = NULL;
#endif

int
dbupdate(DB *db, char *ip, int add, int type)
{
	DBT		dbk, dbd;
	struct gdata	gd;
	time_t		now;
	int		r;
	struct addrinfo hints, *res;

	now = time(NULL);
	memset(&hints, 0, sizeof(hints));
	hints.ai_family = PF_UNSPEC;
	hints.ai_socktype = SOCK_DGRAM;	/*dummy*/
	hints.ai_flags = AI_NUMERICHOST;
	if (add && (type == TRAPHIT || type == WHITE)) {
		if (getaddrinfo(ip, NULL, &hints, &res) != 0) {
			warnx("invalid ip address %s", ip);
			goto bad;
		}
		freeaddrinfo(res);
	}
	memset(&dbk, 0, sizeof(dbk));
	dbk.size = strlen(ip);
	dbk.data = ip;
	memset(&dbd, 0, sizeof(dbd));
	if (!add) {
		/* remove entry */
		r = db->get(db, &dbk, &dbd, 0);
		if (r == -1) {
			warn("db->get failed");
			goto bad;
		}
		if (r) {
			warnx("no entry for %s", ip);
			goto bad;
		} else if (db->del(db, &dbk, 0)) {
			warn("db->del failed");
			goto bad;
		}
	} else {
#ifdef __FreeBSD__
		/* sync the entry to all peers */
		if (syncsend) {
			switch (type) {
			case WHITE:
				sync_white(now, now + WHITEEXP, ip);
				syslog_r(LOG_INFO, &sdata, "send white %s", ip);
				break;
			case TRAPHIT:
				sync_trapped(now, now + TRAPEXP, ip);
				syslog_r(LOG_INFO, &sdata, "send trapped %s", ip);
				break;
			default:
				break;
			}
		}
#endif
		/* add or update entry */
		r = db->get(db, &dbk, &dbd, 0);
		if (r == -1) {
			warn("db->get failed");
			goto bad;
		}
		if (r) {
			int i;

			/* new entry */
			memset(&gd, 0, sizeof(gd));
			gd.first = now;
			gd.bcount = 1;
			switch (type) {
			case WHITE:
				gd.pass = now;
				gd.expire = now + WHITEEXP;
				break;
			case TRAPHIT:
				gd.expire = now + TRAPEXP;
				gd.pcount = -1;
				break;
			case SPAMTRAP:
				gd.expire = 0;
				gd.pcount = -2;
				/* ensure address is lower case*/
				for (i = 0; ip[i] != '\0'; i++)
					if (isupper(ip[i]))
						ip[i] = (char)tolower(ip[i]);
				break;
			default:
				errx(-1, "unknown type %d", type);
			}
			memset(&dbk, 0, sizeof(dbk));
			dbk.size = strlen(ip);
			dbk.data = ip;
			memset(&dbd, 0, sizeof(dbd));
			dbd.size = sizeof(gd);
			dbd.data = &gd;
			r = db->put(db, &dbk, &dbd, 0);
			if (r) {
				warn("db->put failed");
				goto bad;
			}
		} else {
			if (dbd.size != sizeof(gd)) {
				/* whatever this is, it doesn't belong */
				db->del(db, &dbk, 0);
				goto bad;
			}
			memcpy(&gd, dbd.data, sizeof(gd));
			gd.pcount++;
			switch (type) {
			case WHITE:
				gd.pass = now;
				gd.expire = now + WHITEEXP;
				break;
			case TRAPHIT:
				gd.expire = now + TRAPEXP;
				gd.pcount = -1;
				break;
			case SPAMTRAP:
				gd.expire = 0; /* XXX */
				gd.pcount = -2;
				break;
			default:
				errx(-1, "unknown type %d", type);
			}

			memset(&dbk, 0, sizeof(dbk));
			dbk.size = strlen(ip);
			dbk.data = ip;
			memset(&dbd, 0, sizeof(dbd));
			dbd.size = sizeof(gd);
			dbd.data = &gd;
			r = db->put(db, &dbk, &dbd, 0);
			if (r) {
				warn("db->put failed");
				goto bad;
			}
		}
	}
	return (0);
 bad:
	return (1);
}

int
dblist(DB *db)
{
	DBT		dbk, dbd;
	struct gdata	gd;
	int		r;

	/* walk db, list in text format */
	memset(&dbk, 0, sizeof(dbk));
	memset(&dbd, 0, sizeof(dbd));
	for (r = db->seq(db, &dbk, &dbd, R_FIRST); !r;
	    r = db->seq(db, &dbk, &dbd, R_NEXT)) {
		char *a, *cp;

		if ((dbk.size < 1) || dbd.size != sizeof(struct gdata)) {
			db->close(db);
			errx(1, "bogus size db entry - bad db file?");
		}
		memcpy(&gd, dbd.data, sizeof(gd));
		a = malloc(dbk.size + 1);
		if (a == NULL)
			err(1, "malloc");
		memcpy(a, dbk.data, dbk.size);
		a[dbk.size]='\0';
		cp = strchr(a, '\n');
		if (cp == NULL) {
			/* this is a non-greylist entry */
			switch (gd.pcount) {
			case -1: /* spamtrap hit, with expiry time */
				printf("TRAPPED|%s|%d\n", a, gd.expire);
				break;
			case -2: /* spamtrap address */
				printf("SPAMTRAP|%s\n", a);
				break;
			default: /* whitelist */
				printf("WHITE|%s|||%d|%d|%d|%d|%d\n", a,
				    gd.first, gd.pass, gd.expire, gd.bcount,
				    gd.pcount);
				break;
			}
		} else {
			char *helo, *from, *to;

			/* greylist entry */
			*cp = '\0';
			helo = cp + 1;
			from = strchr(helo, '\n');
			if (from == NULL) {
				warnx("No from part in grey key %s", a);
				free(a);
				goto bad;
			}
			*from = '\0';
			from++;
			to = strchr(from, '\n');
			if (to == NULL) {
				/* probably old format - print it the
				 * with an empty HELO field instead 
				 * of erroring out.
				 */			  
				printf("GREY|%s|%s|%s|%s|%d|%d|%d|%d|%d\n",
				    a, "", helo, from, gd.first, gd.pass,
				    gd.expire, gd.bcount, gd.pcount);
			
			} else {
				*to = '\0';
				to++;
				printf("GREY|%s|%s|%s|%s|%d|%d|%d|%d|%d\n",
				    a, helo, from, to, gd.first, gd.pass,
				    gd.expire, gd.bcount, gd.pcount);
			}
		}
		free(a);
	}
	db->close(db);
	db = NULL;
	return (0);
 bad:
	db->close(db);
	db = NULL;
	errx(1, "incorrect db format entry");
	/* NOTREACHED */
	return (1);
}

extern char *__progname;

static int
usage(void)
{
#ifdef __FreeBSD__
	fprintf(stderr, 
		"usage: %s [-D] [-Y synctarget] [[-Tt] -a keys] [[-Tt] -d keys]\n",
		__progname);
#else		
	fprintf(stderr, "usage: %s [[-Tt] -a keys] [[-Tt] -d keys]\n", __progname);
#endif	
	exit(1);
	/* NOTREACHED */
}

#ifdef __FreeBSD__
/* check if parameters -D or -Y applied after -t or -ta parameter */
void
check_opt_sequence(char **argv, int argc)
{
	int i; 
	for (i=0; i<argc; i++)
		if (argv[i][0] != '\0') {
			if (argv[i][1] == 'Y' || argv[i][1] == 'D') {
				fprintf(stderr, 
					"\n%s: wrong opt sequence detected !\n",
					__progname);
				usage();
			}
		}
}
#endif

int
main(int argc, char **argv)
{
	int i, ch, action = 0, type = WHITE, r = 0, c = 0;
	HASHINFO	hashinfo;
	DB		*db;

#ifdef __FreeBSD__
	int syncfd = 0;
	struct servent *ent;
	char *sync_iface = NULL;
	char *sync_baddr = NULL;

	if ((ent = getservbyname("spamd-sync", "udp")) == NULL)
		errx(1, "Can't find service \"spamd-sync\" in /etc/services");
	sync_port = ntohs(ent->s_port);

	while ((ch = getopt(argc, argv, "DY:adtT")) != -1) {
		switch (ch) {
		case 'D':
			debug = 1;
			break;
		case 'Y':
			if (sync_addhost(optarg, sync_port) != 0)
				sync_iface = optarg;
			syncsend++;
			break;
#else			
	while ((ch = getopt(argc, argv, "adtT")) != -1) {
		switch (ch) {
#endif	
		case 'a':
			action = 1;
			break;
		case 'd':
			action = 2;
			break;
		case 't':
			type = TRAPHIT;
			break;
		case 'T':
			type = SPAMTRAP;
			break;
		default:
			usage();
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (action == 0 && type != WHITE)
		usage();
	
	memset(&hashinfo, 0, sizeof(hashinfo));
	db = dbopen(PATH_SPAMD_DB, O_EXLOCK | (action ? O_RDWR : O_RDONLY),
	    0600, DB_HASH, &hashinfo);
	if (db == NULL) {
		if (errno == EFTYPE)	
			err(1, "%s is old, run current spamd to convert it",
			    PATH_SPAMD_DB);
		else 
			err(1, "cannot open %s for %s", PATH_SPAMD_DB,
			    action ? "writing" : "reading");
	}

#ifdef __FreeBSD__
	/* sync only WHITE and TRAPHIT */
	if (action != 1 || type == SPAMTRAP)
		syncsend = 0;
	if (syncsend) {
		openlog_r("spamdb", LOG_NDELAY, LOG_DAEMON, &sdata);
		syncfd = sync_init(sync_iface, sync_baddr, sync_port);
		if (syncfd == -1)
			err(1, "sync init");
	}
#endif	

	switch (action) {
	case 0:
		return dblist(db);
	case 1:
#ifdef __FreeBSD__
		check_opt_sequence(argv, argc); 
#endif		
		for (i=0; i<argc; i++)
			if (argv[i][0] != '\0') {
				c++;
				r += dbupdate(db, argv[i], 1, type);
			}
		if (c == 0)
			errx(2, "no addresses specified");
		break;
	case 2:
#ifdef __FreeBSD__
		check_opt_sequence(argv, argc); 
#endif		
		for (i=0; i<argc; i++)
			if (argv[i][0] != '\0') {
				c++;
				r += dbupdate(db, argv[i], 0, type);
			}
		if (c == 0)
			errx(2, "no addresses specified");
		break;
	default:
		errx(-1, "bad action");
	}
#ifdef __FreeBSD__
	if (syncsend)
		closelog_r(&sdata);
#endif	
	db->close(db);
	return (r);
}
